### ISSUE TITLE
Example infinite_send_to_child.py is unstable

### ISSUE THREAD

**Issue by jgehrcke** [jgehrcke/gipc#49]
Got this in CI run on master: https://travis-ci.org/jgehrcke/gipc/jobs/432268803
```
+timeout --preserve-status --signal=INT --kill-after=7 5 python infinite_send_to_child.py
Child process got message from pipe:
	'written to pipe from a greenlet running in the main process'
Child process got message from pipe:
	'written to pipe from a greenlet running in the main process'
Child process got message from pipe:
	'written to pipe from a greenlet running in the main process'
Child process got message from pipe:
	'written to pipe from a greenlet running in the main process'
Child process got message from pipe:
	'written to pipe from a greenlet running in the main process'
Traceback (most recent call last):
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/libev/corecffi.py", line 404, in _check_callback
    pass
KeyboardInterrupt
Sun Sep 23 22:59:34 2018
Traceback (most recent call last):
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/libev/corecffi.py", line 404, in _check_callback
    pass
KeyboardInterrupt
Sun Sep 23 22:59:34 2018
Traceback (most recent call last):
  File "infinite_send_to_child.py", line 33, in <module>
    main()
  File "infinite_send_to_child.py", line 18, in main
    p.join()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gipc/gipc.py", line 530, in join
    self._returnevent.wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 219, in wait
    return self._wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 129, in _wait
    gotit = self._wait_core(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 106, in _wait_core
    result = self.hub.switch()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/hub.py", line 630, in switch
    return RawGreenlet.switch(self)
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 53, in switch
    return self.__switch('switch', (args, kwds))
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 92, in __switch
    args, kwds = unbound_method(current, *baseargs, to=target)
KeyboardInterrupt
```

**Comment by jgehrcke** [comment]
I hope https://github.com/jgehrcke/gipc/pull/50 was sufficient. But I keep this open for another while.

**Comment by jgehrcke** [comment]
```
Sat Sep 29 13:28:16 2018
Traceback (most recent call last):
  File "infinite_send_to_child.py", line 42, in <module>
    main()
  File "infinite_send_to_child.py", line 19, in main
    p.join()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gipc/gipc.py", line 530, in join
    self._returnevent.wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 219, in wait
    return self._wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 129, in _wait
    gotit = self._wait_core(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 106, in _wait_core
    result = self.hub.switch()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/hub.py", line 630, in switch
    return RawGreenlet.switch(self)
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 53, in switch
    return self.__switch('switch', (args, kwds))
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 92, in __switch
    args, kwds = unbound_method(current, *baseargs, to=target)
KeyboardInterrupt
+result=1
```

and

```
Traceback (most recent call last):
  File "infinite_send_to_child.py", line 42, in <module>
    main()
  File "infinite_send_to_child.py", line 19, in main
    p.join()
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/gipc/gipc.py", line 530, in join
    self._returnevent.wait(timeout)
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/gevent/event.py", line 219, in wait
    return self._wait(timeout)
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/gevent/event.py", line 129, in _wait
    gotit = self._wait_core(timeout)
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/gevent/event.py", line 106, in _wait_core
    result = self.hub.switch()
  File "/home/travis/virtualenv/python3.4.6/lib/python3.4/site-packages/gevent/hub.py", line 630, in switch
    return RawGreenlet.switch(self)
KeyboardInterrupt
```

(seen in #31 -- but I think I have to rebase that one on master first)

**Comment by jgehrcke** [comment]
Does that mean that KeyboardInterrupt during `join()` is not always properly caught despite the `except KeyboardInterrupt`?

```
Traceback (most recent call last):
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/libev/corecffi.py", line 404, in _check_callback
    pass
KeyboardInterrupt
Sat Sep 29 13:28:16 2018
Traceback (most recent call last):
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/libev/corecffi.py", line 404, in _check_callback
    pass
KeyboardInterrupt
Sat Sep 29 13:28:16 2018
Traceback (most recent call last):
  File "infinite_send_to_child.py", line 42, in <module>
    main()
  File "infinite_send_to_child.py", line 19, in main
    p.join()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gipc/gipc.py", line 530, in join
    self._returnevent.wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 219, in wait
    return self._wait(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 129, in _wait
    gotit = self._wait_core(timeout)
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/event.py", line 106, in _wait_core
    result = self.hub.switch()
  File "/home/travis/virtualenv/pypy2.7-5.8.0/site-packages/gevent/hub.py", line 630, in switch
    return RawGreenlet.switch(self)
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 53, in switch
    return self.__switch('switch', (args, kwds))
  File "/opt/python/pypy2.7-5.8.0/lib_pypy/greenlet.py", line 92, in __switch
    args, kwds = unbound_method(current, *baseargs, to=target)
KeyboardInterrupt
```
That's a tough one.

**Comment by jgehrcke** [comment]
Travis build log with the flake: [180929_keyboardinterrupt_problem.log](https://github.com/jgehrcke/gipc/files/2430678/180929_keyboardinterrupt_problem.log)

**Comment by jgehrcke** [comment]
Seen on Python 3.6.6 current master:
```
+timeout --preserve-status --signal=INT --kill-after=7 5 python infinite_send_to_child.py
Child process got message through pipe:
	'I was sent from a greenlet running in the main process!'
Child process got message through pipe:
	'I was sent from a greenlet running in the main process!'
Child process got message through pipe:
	'I was sent from a greenlet running in the main process!'
Child process got message through pipe:
	'I was sent from a greenlet running in the main process!'
Child process got message through pipe:
	'I was sent from a greenlet running in the main process!'
KeyboardInterrupt
Sat Sep 29 15:49:29 2018
KeyboardInterrupt
Sat Sep 29 15:49:29 2018
Traceback (most recent call last):
  File "infinite_send_to_child.py", line 42, in <module>
    main()
  File "infinite_send_to_child.py", line 19, in main
    p.join()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gipc/gipc.py", line 530, in join
    self._returnevent.wait(timeout)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/event.py", line 219, in wait
    return self._wait(timeout)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/event.py", line 129, in _wait
    gotit = self._wait_core(timeout)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/event.py", line 106, in _wait_core
    result = self.hub.switch()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/gevent/hub.py", line 630, in switch
    return RawGreenlet.switch(self)
KeyboardInterrupt
+result=1
```
(CPython 3.6.6, Linux, gevent 1.2.2)

**Comment by jgehrcke** [comment]
Very likely the last test failures happened as of the same instability which is addressed in #53.

**Comment by jgehrcke** [comment]
Improved stability via #54.
